android/src/com/frostwire/android/gui/activities/SettingsActivity2.java

```diff
@@ -147,6 +309,82 @@ public void onCreate(Bundle savedInstanceState) {
+                    if (!cbPreference.isChecked()) {
+                        selectAll.setChecked(false);
+                        if (areAllEnginesChecked(false, activeSearchEnginePreferences)) {
+                            cbPreference.setChecked(true); // always keep one checked
+                        }
+                    } else {
+                        boolean allChecked = areAllEnginesChecked(true, activeSearchEnginePreferences);
+                        selectAll.setChecked(allChecked);
+                    }
+                    selectAll.setClickListenerEnabled(true);
+                    return true;
+                }
+            };
+
+            // Hide inactive search engines and setup click listeners to interact with Select All.
+            if (searchEnginesScreen != null) {
+                for (CheckBoxPreference preference : inactiveSearchPreferences.keySet()) {
+                    searchEnginesScreen.removePreference(preference);
+                }
+            }
+
+            for (CheckBoxPreference preference : activeSearchEnginePreferences.keySet()) {
+                preference.setOnPreferenceClickListener(searchEngineClickListener);
+            }
+
+            ToggleAllSearchEnginesPreference2 selectAll = (ToggleAllSearchEnginesPreference2) findPreference("frostwire.prefs.search.preference_category.select_all");
+            selectAll.setSearchEnginePreferences(activeSearchEnginePreferences);
+        }
+
+        private boolean areAllEnginesChecked(boolean checked, Map<CheckBoxPreference, SearchEngine> activeSearchEnginePreferences) {
+            final Collection<CheckBoxPreference> preferences = activeSearchEnginePreferences.keySet();
+            for (CheckBoxPreference preference : preferences) {
+                if (checked != preference.isChecked()) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private void getSearchEnginePreferences(Map<CheckBoxPreference, SearchEngine> inactiveSearchEnginePreferences, Map<CheckBoxPreference, SearchEngine> activeSearchEnginePreferences) {
+            // make sure we start empty
+            inactiveSearchEnginePreferences.clear();
+            activeSearchEnginePreferences.clear();
+
+            for (SearchEngine engine : SearchEngine.getEngines()) {
+                CheckBoxPreference preference = (CheckBoxPreference) findPreference(engine.getPreferenceKey());
+                if (preference != null) { //it could already have been removed due to remote config value.
+                    if (engine.isActive()) {
+                        activeSearchEnginePreferences.put(preference, engine);
+                    } else {
+                        inactiveSearchEnginePreferences.put(preference, engine);
+                    }
+                }
+            }
+        }
     }
 
     public static class Torrent extends PreferenceFragment {
```

Comment:
the name of this class is very misleading. It should be more like TorrentPreferenceFragment